packages/beacon-state-transition/src/allForks/util/weakSubjectivity.ts

```diff
@@ -15,3 +31,69 @@ export function getLatestWeakSubjectivityCheckpointEpoch(
   const valCount = state.epochCtx.currentShuffling.activeIndices.length;
   return getWeakSubjectivityCheckpointEpoch(config, state.finalizedCheckpoint.epoch, valCount, safetyDecay);
 }
+
+/**
+  Returns the weak subjectivity period for the current ``state``. 
+    This computation takes into account the effect of:
+      - validator set churn (bounded by ``get_validator_churn_limit()`` per epoch), and 
+      - validator balance top-ups (bounded by ``MAX_DEPOSITS * SLOTS_PER_EPOCH`` per epoch).
+    A detailed calculation can be found at:
+    https://github.com/runtimeverification/beacon-chain-verification/blob/master/weak-subjectivity/weak-subjectivity-analysis.pdf
+ */
+export function computeWeakSubjectivityPeriod(config: IBeaconConfig, state: allForks.BeaconState): number {
+  let wsPeriod = config.MIN_VALIDATOR_WITHDRAWABILITY_DELAY;
+  const ethToGwei = Number(ETH_TO_GWEI);
+  const currentEpoch = getCurrentEpoch(state);
+  const indices = getActiveValidatorIndices(state, currentEpoch);
+  const N = indices.length;
+  const totalBalance = getTotalBalance(state, indices);
+  const t = Math.floor(Number(totalBalance) / N / ethToGwei);
+  const T = Math.floor(Number(MAX_EFFECTIVE_BALANCE) / ethToGwei);
+  const delta = getChurnLimit(config, N);
+  // eslint-disable-next-line @typescript-eslint/naming-convention
+  const Delta = MAX_DEPOSITS * SLOTS_PER_EPOCH;
+  const D = Number(SAFETY_DECAY);
+
+  if (T * (200 + 3 * D) < t * (200 + 12 * D)) {
+    const epochsForValidatorSetChurn = Math.floor(
+      (N * (t * (200 + 12 * D) - T * (200 + 3 * D))) / (600 * delta * (2 * t + T))
+    );
+    const epochsForBalanceTopUps = Math.floor((N * (200 + 3 * D)) / (600 * Delta));
+    wsPeriod +=
+      epochsForValidatorSetChurn > epochsForBalanceTopUps ? epochsForValidatorSetChurn : epochsForBalanceTopUps;
+  } else {
+    wsPeriod += Math.floor((3 * N * D * t) / (200 * Delta * (T - t)));
+  }
+  return wsPeriod;
+}
+
+function getLatestBlockRoot(config: IBeaconConfig, state: allForks.BeaconState): Root {
+  const header = ssz.phase0.BeaconBlockHeader.clone(state.latestBlockHeader);
+  if (ssz.Root.equals(header.stateRoot, ZERO_HASH)) {
+    header.stateRoot = config.getForkTypes(state.slot).BeaconState.hashTreeRoot(state);
+  }
+  return ssz.phase0.BeaconBlockHeader.hashTreeRoot(header);
+}
+
+export function isWithinWeakSubjectivityPeriod(
+  config: IBeaconConfig,
+  genesisTime: number,
+  wsState: allForks.BeaconState,
+  wsCheckpoint?: Checkpoint
```

Comment:
It's optional because in the case that the user doesn't provide a wsCheckpoint, we are implicitly trusting the state that we get from Infura.  This may not be ideal, but if the user does not provide us a block root + epoch, we need to fetch the data for the latest block root + epoch that we intend to fetch because it seems like it would be pointless to verify the wsState against data that is inside the wsState, such as the checkpoint inside the wsState, at least inasmuch as we are trying to reinforce some security on the fetched Infura state.

Reply:
I think we should keep the param as non-optional just to simplify the code. If no wsCheckpoint is provided via cli, we can just create a checkpoint from the ws state to pass here and to perform the safety checks.